tests/test_oic_provider_logout.py

```diff
@@ -0,0 +1,831 @@
+        ],
+    },
+    "a1b2c3": {
+        "redirect_uris": [("http://localhost:8087/authz", None)],
+        "client_salt": "salted",
+        "client_secret": "very_secret",
+        "response_types": ["code", "token", "code id_token"],
+    },
+    "client0": {
+        "redirect_uris": [("http://www.example.org/authz", None)],
+        "client_secret": "very_secret",
+        "post_logout_redirect_uris": [("https://www.example.org/post_logout", None)],
+        "client_salt": "salted",
+        "response_types": ["code", "token", "code id_token"],
+    },
+    CLIENT_ID: {
+        "client_secret": CLIENT_SECRET,
+        "redirect_uris": [("http://localhost:8087/authz", None)],
+        "client_salt": "salted",
+        "token_endpoint_auth_method": "client_secret_post",
+        "response_types": ["code", "token", "code id_token"],
+    },
+}  # type: Dict[str, Dict[str, Any]]
+
+USERDB = {
+    "user": {
+        "name": "Hans Granberg",
+        "nickname": "Hasse",
+        "email": "hans@example.org",
+        "verified": False,
+        "sub": "user",
+    },
+    "username": {
+        "name": "Linda Lindgren",
+        "nickname": "Linda",
+        "email": "linda@example.com",
+        "verified": True,
+        "sub": "username",
+        "extra_claim": "extra_claim_value",
+    },
+}
+
+URLMAP = {CLIENT_ID: ["https://example.com/authz"]}
+
+
+def dummy_template_renderer(template_name: str, context: Dict[str, str]) -> str:
+    if template_name == "verify_logout":
+        body = ["<dl>"]
+        body.extend(["<dt>{}</dt><dd>{}</dd>".format(k, v) for k, v in context.items()])
+        body.append("</dl>")
+        _html = """
+<html lang="en" xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
+<head>
+  <title>Verify Logout</title>
+</head>
+<body>
+    {}
+</body>
+</html>
+        """.format(
+            "\n".join(body)
+        )
+        return _html
+    else:
+        return ""
+
+
+class DummyEventStore(object):
+    def __init__(self):
+        self.db = {}  # type: Dict[str, str]
+
+    def store(self, typ, val):
+        self.db[typ] = val
+
+
+def _eq(l1, l2):
+    return set(l1) == set(l2)
+
+
+class DummyAuthn(UserAuthnMethod):
+    def __init__(self, srv, user):
+        UserAuthnMethod.__init__(self, srv)
+        self.user = user
+
+    def authenticated_as(self, cookie=None, **kwargs):
+        if cookie == "FAIL":
+            return None, 0
+        else:
+            return {"uid": self.user}, time()
+
+
+AUTHN_BROKER = AuthnBroker()
+AUTHN_BROKER.add("UNDEFINED", DummyAuthn(None, "username"))
+
+# dealing with authorization
+AUTHZ = AuthzHandling()
+SYMKEY = rndstr(16)  # symmetric key used to encrypt cookie info
+USERINFO = UserInfo(USERDB)
+
+
+class DummyHTTPClient(object):
+    def __init__(self, resp_status=200):
+        self.resp_status = resp_status
+
+    def http_request(self, *args, **kwargs):
+        return Response(status_code=self.resp_status)
+
+
+class TestProvider(object):
```

Comment:
I try to avoid using pytest in favor of unittest. The are (in my opinion) better and provide handy assert functions so you don't have to write `assert` statements.

I won't enforce this, but please consider it.

Reply:
Haven't used unittest these last 6-7 years. Switched to py.test because it was so much better. That might not be the case anymore ?

Reply:
I believe they got much better and they are in standard library so that is a plus :0